lib/TPP/ConvInitSimplify.cpp

```diff
@@ -0,0 +1,113 @@
+//===- FastMathOpts.cpp.cpp --------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Tpp/TppUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+static bool isBroadCastOp(linalg::GenericOp linalgOp) {
+  if (linalgOp->getNumOperands() != 2 || linalgOp->getNumResults() != 1)
+    return false;
+  if (!tpp::utils::hasOnlyYieldOp(linalgOp.getRegion()))
+    return false;
+  SmallVector<unsigned> perm;
+  return linalgOp.getMatchingIndexingMap(linalgOp.getDpsInputOperand(0))
+      .isPermutationOfMinorIdentityWithBroadcasting(perm);
+}
+
+static std::optional<linalg::GenericOp> getBroadCastProdcuer(OpOperand *rhs) {
+  linalg::GenericOp broadcastOp = rhs->get().getDefiningOp<linalg::GenericOp>();
+  if (!broadcastOp || !isBroadCastOp(broadcastOp))
+    return std::nullopt;
+  return broadcastOp;
+}
+
+static bool isConstantZero(linalg::FillOp fillOp) {
+  Value cst = fillOp.getInputs()[0];
+  return matchPattern(cst, m_AnyZeroFloat()) || matchPattern(cst, m_Zero());
+}
+
+static bool isFillWithZeros(linalg::Conv2DNhwcHwcfOp convOp) {
+  auto fillOp =
+      convOp.getDpsInitOperand(0)->get().getDefiningOp<linalg::FillOp>();
+  if (!fillOp)
+    return false;
+  return isConstantZero(fillOp);
+}
+
+// Instead of initializing the output of a convolution with zero and then add a
+// bias, initialize the output of the convolution with the bias.
+struct ConvInitSimplifyImpl : public OpRewritePattern<linalg::GenericOp> {
```

Comment:
What is your plan on moving the `addOp` above the `ConvOp`?

If you're going to create a new rewrite pattern (I recommend this), then this one needs to be called by what it does: "eliminate zero init and an add to bias init".

If you'll bundle the "add after relu" here, then it can continue with the name it is. But then the rewrite code will be a bit messy once we start adding more transforms...

Reply:
I don't have a plan. But I suppose I can add more logic here when I have a test case, or re-evaluate naming later on.

Reply:
I'd strongly recommend you create two different patterns, definitely reusing code to detect linalg+add etc, but with different names. So for now, just rename this to what it actually does.